habitat_baselines/il/trainers/pacman_trainer.py

```diff
@@ -0,0 +1,654 @@
+                (10, 15),
+                font,
+                scale,
+                color,
+                thickness,
+            )
+
+            frames_with_text.append(frame)
+
+        images_to_video(frames_with_text, results_dir, "ep_" + str(t))
+
+    def train(self) -> None:
+        r"""Main method for training Navigation model of EQA.
+
+        Returns:
+            None
+        """
+        config = self.config
+
+        with habitat.Env(config.TASK_CONFIG) as env:
+            nav_dataset = (
+                NavDataset(
+                    config,
+                    env,
+                    self.device,
+                )
+                .shuffle(1000)
+                .decode("rgb")
+            )
+
+            nav_dataset = nav_dataset.map(nav_dataset.map_dataset_sample)
+
+            train_loader = DataLoader(
+                nav_dataset, batch_size=config.IL.NAV.batch_size
+            )
+
+            logger.info("train_loader has {} samples".format(len(nav_dataset)))
+
+            q_vocab_dict, _ = nav_dataset.get_vocab_dicts()
+
+            model_kwargs = {"q_vocab": q_vocab_dict.word2idx_dict}
+            model = NavPlannerControllerModel(**model_kwargs)
+
+            planner_loss_fn = MaskedNLLCriterion()
+            controller_loss_fn = MaskedNLLCriterion()
+
+            optim = torch.optim.Adam(
+                filter(lambda p: p.requires_grad, model.parameters()),
+                lr=float(config.IL.NAV.lr),
+            )
+
+            metrics = NavMetric(
+                info={"split": "train"},
+                metric_names=["planner_loss", "controller_loss"],
+                log_json=os.path.join(config.OUTPUT_LOG_DIR, "train.json"),
+            )
+
+            epoch = 1
+
+            avg_p_loss = 0.0
+            avg_c_loss = 0.0
+
+            logger.info(model)
+            model.train().to(self.device)
+
+            with TensorboardWriter(
+                config.TENSORBOARD_DIR, flush_secs=self.flush_secs
+            ) as writer:
+                while epoch <= config.IL.NAV.max_epochs:
+                    start_time = time.time()
+                    for t, batch in enumerate(train_loader):
+                        (
+                            idx,
+                            questions,
+                            _,
+                            planner_img_feats,
+                            planner_actions_in,
+                            planner_actions_out,
+                            planner_action_lengths,
+                            planner_masks,
+                            controller_img_feats,
+                            controller_actions_in,
+                            planner_hidden_idx,
+                            controller_outs,
+                            controller_action_lengths,
+                            controller_masks,
+                        ) = batch
+
+                        questions = questions.to(self.device)
```

Comment:
Why don't you call .to while the batch is still in a Tuple form? Would be a lot cleaner. Also can add `non_blocking=True`